Coverage commands in RBE crash Bazel

### Description of the bug:


I'm trying to add regression testing for generating coverage reports for Rust in RBE environments on https://github.com/bazelbuild/rules_rust/pull/2005 and I run into the following crash:

https://buildkite.com/bazel/rules-rust-rustlang/builds/10127#018c793a-1630-4729-a194-396fb371e6bf
```
04:45:40) ERROR: <builtin>: Coverage report generation failed: (Exit 34): INVALID_ARGUMENT: Invalid arguments:
  "command.ValidateSpec": Invalid spec - docker container must be specified
java.io.IOException: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Invalid arguments:
  "command.ValidateSpec": Invalid spec - docker container must be specified
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.executeRemotely(GrpcRemoteExecutor.java:241)
	at com.google.devtools.build.lib.remote.RemoteExecutionService.executeRemotely(RemoteExecutionService.java:1493)
	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.lambda$exec$2(RemoteSpawnRunner.java:292)
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:245)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:127)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:116)
	at com.google.devtools.build.lib.remote.RemoteSpawnRunner.exec(RemoteSpawnRunner.java:265)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:159)
	at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:119)
	at com.google.devtools.build.lib.exec.SpawnStrategyResolver.exec(SpawnStrategyResolver.java:45)
	at com.google.devtools.build.lib.bazel.coverage.CoverageReportActionBuilder$CoverageReportAction.execute(CoverageReportActionBuilder.java:140)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1148)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1065)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
	at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
	at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:562)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:859)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:333)
	at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:171)
	at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
	at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: io.grpc.StatusRuntimeException: INVALID_ARGUMENT: Invalid arguments:
  "command.ValidateSpec": Invalid spec - docker container must be specified
	at io.grpc.Status.asRuntimeException(Status.java:535)
	at io.grpc.stub.ClientCalls$BlockingResponseStream.hasNext(ClientCalls.java:660)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$2(GrpcRemoteExecutor.java:175)
	at com.google.devtools.build.lib.remote.Retrier.execute(Retrier.java:245)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:127)
	at com.google.devtools.build.lib.remote.RemoteRetrier.execute(RemoteRetrier.java:116)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.lambda$executeRemotely$3(GrpcRemoteExecutor.java:146)
	at com.google.devtools.build.lib.remote.util.Utils.refreshIfUnauthenticated(Utils.java:525)
	at com.google.devtools.build.lib.remote.GrpcRemoteExecutor.executeRemotely(GrpcRemoteExecutor.java:144)
	... 26 more
```

### Which category does this issue belong to?


Remote Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.


Use the changes at https://github.com/bazelbuild/rules_rust/pull/2005 to build with RBE as described by presubmit.yaml

```yaml
rbe_ubuntu2004:
    shell_commands:
      - sed -i 's/^# load("@bazelci_rules/load("@bazelci_rules/' WORKSPACE.bazel
      - sed -i 's/^# rbe_preconfig/rbe_preconfig/' WORKSPACE.bazel
    coverage_targets:
      - "--"
      - "//..."
```

### Which operating system are you running Bazel on?


Linux

### What is the output of `bazel info release`?


7.0.0

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.


_No response_

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?


_No response_

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.


_No response_

### Have you found anything relevant by searching the web?

_No response_

### Any other information, logs, or outputs that you want to share?

_No response_

--- a/src/main/java/com/google/devtools/build/lib/analysis/test/TestActionBuilder.java
+++ b/src/main/java/com/google/devtools/build/lib/analysis/test/TestActionBuilder.java
@@ -39,6 +39,7 @@ import com.google.devtools.build.lib.analysis.TransitiveInfoCollection;
 import com.google.devtools.build.lib.analysis.actions.LazyWriteNestedSetOfTupleAction;
 import com.google.devtools.build.lib.analysis.config.BuildConfigurationValue;
 import com.google.devtools.build.lib.analysis.test.TestProvider.TestParams;
+import com.google.devtools.build.lib.analysis.test.TestProvider.TestParams.CoverageParams;
 import com.google.devtools.build.lib.collect.nestedset.NestedSet;
 import com.google.devtools.build.lib.collect.nestedset.NestedSetBuilder;
 import com.google.devtools.build.lib.collect.nestedset.Order;
@@ -53,9 +54,7 @@ import java.util.TreeMap;
 import java.util.TreeSet;
 import javax.annotation.Nullable;
 
-/**
- * Helper class to create test actions.
- */
+/** Helper class to create test actions. */
 public final class TestActionBuilder {
   private static final String CC_CODE_COVERAGE_SCRIPT = "CC_CODE_COVERAGE_SCRIPT";
   private static final String LCOV_MERGER = "LCOV_MERGER";
@@ -98,8 +97,7 @@ public final class TestActionBuilder {
         "invalid",
         ImmutableList.of(),
         ImmutableList.of(),
-        FilesToRunProvider.EMPTY,
-        ImmutableList.of());
+        /* coverageParams= */ null);
   }
 
   /**
@@ -226,8 +224,7 @@ public final class TestActionBuilder {
         new TestTargetProperties(ruleContext, executionRequirements);
 
     // If the test rule does not provide InstrumentedFilesProvider, there's not much that we can do.
-    final boolean collectCodeCoverage = config.isCodeCoverageEnabled()
-        && instrumentedFiles != null;
+    final boolean collectCodeCoverage = config.isCodeCoverageEnabled() && instrumentedFiles != null;
 
     Artifact testActionExecutable =
         isUsingTestWrapperInsteadOfTestSetupScript
@@ -329,8 +326,8 @@ public final class TestActionBuilder {
       }
 
       Artifact instrumentedFileManifest =
-          InstrumentedFileManifestAction.getInstrumentedFileManifest(ruleContext,
-              instrumentedFiles.getInstrumentedFiles(), metadataFiles);
+          InstrumentedFileManifestAction.getInstrumentedFileManifest(
+              ruleContext, instrumentedFiles.getInstrumentedFiles(), metadataFiles);
       executionSettings =
           new TestTargetExecutionSettings(
               ruleContext,
@@ -456,18 +453,20 @@ public final class TestActionBuilder {
         results.add(cacheStatus);
       }
     }
-    // TODO(bazel-team): Passing the reportGenerator to every TestParams is a bit strange.
-    FilesToRunProvider reportGenerator = null;
+    CoverageParams coverageParams = null;
     if (config.isCodeCoverageEnabled()) {
+      // TODO(bazel-team): Passing the reportGenerator to every TestParams is a bit strange.
       // It's not enough to add this if the rule has coverage enabled because the command line may
       // contain rules with baseline coverage but no test rules that have coverage enabled, and in
       // that case, we still need the report generator.
       TransitiveInfoCollection reportGeneratorTarget =
           ruleContext.getPrerequisite(":coverage_report_generator");
-      reportGenerator = reportGeneratorTarget.getProvider(FilesToRunProvider.class);
+      FilesToRunProvider reportGenerator =
+          reportGeneratorTarget.getProvider(FilesToRunProvider.class);
       if (reportGenerator.getExecutable() == null) {
         ruleContext.ruleError("--coverage_report_generator does not refer to an executable target");
       }
+      coverageParams = new CoverageParams(coverageArtifacts.build(), reportGenerator, actionOwner);
     }
 
     return new TestParams(
@@ -477,8 +476,7 @@ public final class TestActionBuilder {
         TestTimeout.getTestTimeout(ruleContext.getRule()),
         ruleContext.getRule().getRuleClass(),
         ImmutableList.copyOf(results),
-        coverageArtifacts.build(),
-        reportGenerator,
-        testOutputs.build());
+        testOutputs.build(),
+        coverageParams);
   }
 }

--- a/src/main/java/com/google/devtools/build/lib/analysis/test/TestProvider.java
+++ b/src/main/java/com/google/devtools/build/lib/analysis/test/TestProvider.java
@@ -16,12 +16,15 @@ package com.google.devtools.build.lib.analysis.test;
 
 import com.google.common.collect.ImmutableList;
 import com.google.devtools.build.lib.actions.ActionInput;
+import com.google.devtools.build.lib.actions.ActionOwner;
 import com.google.devtools.build.lib.actions.Artifact;
 import com.google.devtools.build.lib.analysis.FilesToRunProvider;
 import com.google.devtools.build.lib.analysis.TransitiveInfoCollection;
 import com.google.devtools.build.lib.analysis.TransitiveInfoProvider;
 import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
 import com.google.devtools.build.lib.packages.TestTimeout;
+import com.google.devtools.build.lib.skyframe.serialization.autocodec.AutoCodec;
+import javax.annotation.Nullable;
 
 /** A {@link TransitiveInfoProvider} for configured targets that implement test rules. */
 @Immutable
@@ -54,15 +57,21 @@ public final class TestProvider implements TransitiveInfoProvider {
   /** A value class describing the properties of a test. */
   // Non-final only for mocking.
   public static class TestParams {
+    /** A value class describing the coverage-related properties of a test. */
+    @AutoCodec
+    record CoverageParams(
+        ImmutableList<Artifact> coverageArtifacts,
+        FilesToRunProvider coverageReportGenerator,
+        ActionOwner actionOwner) {}
+
     private final int runs;
     private final int shards;
     private final boolean runsDetectsFlakes;
     private final TestTimeout timeout;
     private final String testRuleClass;
     private final ImmutableList<Artifact.DerivedArtifact> testStatusArtifacts;
-    private final ImmutableList<Artifact> coverageArtifacts;
-    private final FilesToRunProvider coverageReportGenerator;
     private final ImmutableList<ActionInput> outputs;
+    @Nullable private final CoverageParams coverageParams;
 
     /**
      * Don't call this directly. Instead use {@link
@@ -75,18 +84,16 @@ public final class TestProvider implements TransitiveInfoProvider {
         TestTimeout timeout,
         String testRuleClass,
         ImmutableList<Artifact.DerivedArtifact> testStatusArtifacts,
-        ImmutableList<Artifact> coverageArtifacts,
-        FilesToRunProvider coverageReportGenerator,
-        ImmutableList<ActionInput> outputs) {
+        ImmutableList<ActionInput> outputs,
+        @Nullable CoverageParams coverageParams) {
       this.runs = runs;
       this.shards = shards;
       this.runsDetectsFlakes = runsDetectsFlakes;
       this.timeout = timeout;
       this.testRuleClass = testRuleClass;
       this.testStatusArtifacts = testStatusArtifacts;
-      this.coverageArtifacts = coverageArtifacts;
-      this.coverageReportGenerator = coverageReportGenerator;
       this.outputs = outputs;
+      this.coverageParams = coverageParams;
     }
 
     /** Returns the number of times this test should be run. */
@@ -122,19 +129,34 @@ public final class TestProvider implements TransitiveInfoProvider {
       return testStatusArtifacts;
     }
 
-    /** Returns the coverageArtifacts. */
-    public ImmutableList<Artifact> getCoverageArtifacts() {
-      return coverageArtifacts;
-    }
-
-    /** Returns the coverage report generator tool. */
-    public FilesToRunProvider getCoverageReportGenerator() {
-      return coverageReportGenerator;
-    }
-
     /** Returns the list of mandatory and optional test outputs. */
     public ImmutableList<ActionInput> getOutputs() {
       return outputs;
+    }
+
+    /** Returns the coverageArtifacts. */
+    public ImmutableList<Artifact> getCoverageArtifacts() {
+      return coverageParams != null ? coverageParams.coverageArtifacts() : ImmutableList.of();
+    }
+
+    /**
+     * Returns the coverage report generator tool.
+     *
+     * <p>Returns a non-null value if and only iff coverage is generally enabled.
+     */
+    @Nullable
+    public FilesToRunProvider getCoverageReportGenerator() {
+      return coverageParams != null ? coverageParams.coverageReportGenerator() : null;
+    }
+
+    /**
+     * Returns the test action owner.
+     *
+     * <p>Returns a non-null value if and only iff coverage is generally enabled.
+     */
+    @Nullable
+    public ActionOwner getActionOwnerForCoverage() {
+      return coverageParams != null ? coverageParams.actionOwner() : null;
     }
   }
 }

--- a/src/main/java/com/google/devtools/build/lib/bazel/coverage/CoverageArgs.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/coverage/CoverageArgs.java
@@ -17,6 +17,7 @@ package com.google.devtools.build.lib.bazel.coverage;
 import static java.util.Objects.requireNonNull;
 
 import com.google.common.collect.ImmutableList;
+import com.google.devtools.build.lib.actions.ActionOwner;
 import com.google.devtools.build.lib.actions.Artifact;
 import com.google.devtools.build.lib.actions.ArtifactFactory;
 import com.google.devtools.build.lib.actions.ArtifactOwner;
@@ -41,6 +42,7 @@ public record CoverageArgs(
     FilesToRunProvider reportGenerator,
     String workspaceName,
     boolean htmlReport,
+    ActionOwner actionOwner,
     @Nullable PathFragment coverageDir,
     @Nullable Artifact lcovOutput) {
   public CoverageArgs {
@@ -51,6 +53,7 @@ public record CoverageArgs(
     requireNonNull(artifactOwner, "artifactOwner");
     requireNonNull(reportGenerator, "reportGenerator");
     requireNonNull(workspaceName, "workspaceName");
+    requireNonNull(actionOwner, "actionOwner");
   }
 
   public static CoverageArgs create(
@@ -61,7 +64,8 @@ public record CoverageArgs(
       ArtifactOwner artifactOwner,
       FilesToRunProvider reportGenerator,
       String workspaceName,
-      boolean htmlReport) {
+      boolean htmlReport,
+      ActionOwner actionOwner) {
     return new CoverageArgs(
         directories,
         coverageArtifacts,
@@ -71,14 +75,13 @@ public record CoverageArgs(
         reportGenerator,
         workspaceName,
         htmlReport,
+        actionOwner,
         /* coverageDir= */ null,
         /* lcovOutput= */ null);
   }
 
   public static CoverageArgs createCopyWithCoverageDirAndLcovOutput(
-      CoverageArgs args,
-      PathFragment coverageDir,
-      Artifact lcovOutput) {
+      CoverageArgs args, PathFragment coverageDir, Artifact lcovOutput) {
     return new CoverageArgs(
         args.directories(),
         args.coverageArtifacts(),
@@ -88,6 +91,7 @@ public record CoverageArgs(
         args.reportGenerator(),
         args.workspaceName(),
         args.htmlReport(),
+        args.actionOwner(),
         coverageDir,
         lcovOutput);
   }

--- a/src/main/java/com/google/devtools/build/lib/bazel/coverage/CoverageReportActionBuilder.java
+++ b/src/main/java/com/google/devtools/build/lib/bazel/coverage/CoverageReportActionBuilder.java
@@ -14,6 +14,9 @@
 
 package com.google.devtools.build.lib.bazel.coverage;
 
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.primitives.Booleans.falseFirst;
+import static java.util.Comparator.comparing;
 
 import com.google.common.base.Joiner;
 import com.google.common.collect.ImmutableList;
@@ -64,6 +67,7 @@ import com.google.devtools.build.lib.vfs.Path;
 import com.google.devtools.build.lib.vfs.PathFragment;
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Comparator;
 import java.util.List;
 import javax.annotation.Nullable;
 
@@ -98,7 +102,11 @@ public final class CoverageReportActionBuilder {
   private static final ResourceSet LOCAL_RESOURCES =
       ResourceSet.createWithRamCpu(/* memoryMb= */ 750, /* cpu= */ 1);
 
-  private static final ActionOwner ACTION_OWNER = ActionOwner.SYSTEM_ACTION_OWNER;
+  private static final Comparator<ActionOwner> ACTION_OWNER_COMPARATOR =
+      comparing(
+              (ActionOwner actionOwner) -> actionOwner.getExecProperties().isEmpty(), falseFirst())
+          .thenComparing(ActionOwner::getLabel)
+          .thenComparing(ActionOwner::getConfigurationChecksum);
 
   // SpawnActions can't be used because they need the AnalysisEnvironment and this action is
   // created specially at the very end of the analysis phase when we don't have it anymore.
@@ -182,8 +190,7 @@ public final class CoverageReportActionBuilder {
     }
   }
 
-  public CoverageReportActionBuilder() {
-  }
+  public CoverageReportActionBuilder() {}
 
   /** Returns the coverage report action. May return null in case of an error. */
   @Nullable
@@ -205,6 +212,7 @@ public final class CoverageReportActionBuilder {
     }
     ImmutableList.Builder<Artifact> builder = ImmutableList.builder();
     FilesToRunProvider reportGenerator = null;
+    ActionOwner actionOwner = null;
     for (ConfiguredTarget target : targetsToTest) {
       // Skip incompatible tests.
       if (target.get(IncompatiblePlatformProvider.PROVIDER) != null) {
@@ -212,26 +220,45 @@ public final class CoverageReportActionBuilder {
       }
       TestParams testParams = target.getProvider(TestProvider.class).getTestParams();
       builder.addAll(testParams.getCoverageArtifacts());
-      if (reportGenerator == null) {
+      // targetsToTest has non-deterministic order, so we ensure that we pick the same action owner
+      // and matching report generator each time by picking the owner that's lexicographically
+      // largest. We prefer an owner with exec properties set in case the action is run remotely.
+      if (reportGenerator == null
+          || ACTION_OWNER_COMPARATOR.compare(testParams.getActionOwnerForCoverage(), actionOwner)
+              > 0) {
         reportGenerator = testParams.getCoverageReportGenerator();
+        actionOwner = testParams.getActionOwnerForCoverage();
       }
     }
     // If all tests are incompatible, there's nothing to do.
     if (reportGenerator == null) {
       return null;
     }
+    checkNotNull(actionOwner);
     ImmutableList<Artifact> coverageArtifacts = builder.addAll(baselineCoverageArtifacts).build();
     if (!coverageArtifacts.isEmpty()) {
       PathFragment coverageDir = TestRunnerAction.COVERAGE_TMP_ROOT;
-      Artifact lcovArtifact = factory.getDerivedArtifact(
-          coverageDir.getRelative("lcov_files.tmp"),
-          directories.getBuildDataDirectory(workspaceName),
-          artifactOwner);
-      Action lcovFileAction = generateLcovFileWriteAction(lcovArtifact, coverageArtifacts);
-      Action coverageReportAction = generateCoverageReportAction(
-          CoverageArgs.create(directories, coverageArtifacts, lcovArtifact, factory, artifactOwner,
-              reportGenerator, workspaceName, htmlReport),
-          argsFunction, locationFunc);
+      Artifact lcovArtifact =
+          factory.getDerivedArtifact(
+              coverageDir.getRelative("lcov_files.tmp"),
+              directories.getBuildDataDirectory(workspaceName),
+              artifactOwner);
+      Action lcovFileAction =
+          generateLcovFileWriteAction(lcovArtifact, coverageArtifacts, actionOwner);
+      Action coverageReportAction =
+          generateCoverageReportAction(
+              CoverageArgs.create(
+                  directories,
+                  coverageArtifacts,
+                  lcovArtifact,
+                  factory,
+                  artifactOwner,
+                  reportGenerator,
+                  workspaceName,
+                  htmlReport,
+                  actionOwner),
+              argsFunction,
+              locationFunc);
       return new CoverageReportActionsWrapper(
           lcovFileAction, coverageReportAction, actionKeyContext);
     } else {
@@ -242,13 +269,13 @@ public final class CoverageReportActionBuilder {
   }
 
   private static FileWriteAction generateLcovFileWriteAction(
-      Artifact lcovArtifact, ImmutableList<Artifact> coverageArtifacts) {
+      Artifact lcovArtifact, ImmutableList<Artifact> coverageArtifacts, ActionOwner actionOwner) {
     List<String> filepaths = new ArrayList<>(coverageArtifacts.size());
     for (Artifact artifact : coverageArtifacts) {
       filepaths.add(artifact.getExecPathString());
     }
     return FileWriteAction.create(
-        ACTION_OWNER,
+        actionOwner,
         lcovArtifact,
         Joiner.on('\n').join(filepaths),
         /* makeExecutable= */ false,
@@ -291,7 +318,7 @@ public final class CoverageReportActionBuilder {
       inputsBuilder.add(runfilesMiddleman);
     }
     return new CoverageReportAction(
-        ACTION_OWNER,
+        args.actionOwner(),
         inputsBuilder.build(),
         ImmutableSet.of(lcovOutput),
         actionArgs,

